src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+        self.linear_3 = nn.Linear(
+            config.hidden_size // 2,
+            config.text_intermediate_size // 2,
+            bias=False,
+        )
+        self.linear_2 = nn.Linear(
+            config.text_intermediate_size // 2,
+            config.text_hidden_size,
+            bias=False,
+        )
+
+    def forward(self, image_features):
+        hidden_states = self.act(self.linear_1(image_features)) * self.linear_3(image_features)
+        hidden_states = self.linear_2(hidden_states)
+        return hidden_states
+
+
+# Molmo image components inherited from CLIPVision
+# We have different attention classes for the txt and the image components, they need to be propagated back correctly
+
+
+class MolmoVisionAttention(CLIPAttention):
+    pass
+
+
+class MolmoVisionSdpaAttention(MolmoVisionAttention, CLIPSdpaAttention):
+    pass
+
+
+class MolmoVisionFlashAttention2(MolmoVisionAttention, CLIPFlashAttention2):
+    pass
+
+
+MOLMO_VISION_ATTENTION_CLASSES = {
+    "eager": MolmoVisionAttention,
+    "sdpa": MolmoVisionSdpaAttention,
+    "flash_attention_2": MolmoVisionFlashAttention2,
+}
+
+
+class MolmoVisionEmbeddings(nn.Module):
+    def __init__(self, config: MolmoVisionConfig):
+        super().__init__()
+        self.config = config
+        self.embed_dim = config.hidden_size
+        self.image_size = config.image_size
+        self.patch_size = config.patch_size
+
+        self.class_embedding = nn.Parameter(torch.randn(self.embed_dim))
+        self.patch_embedding = nn.Linear(
+            self.patch_size**2 * 3,
+            self.embed_dim,
+            bias=False,
+        )
+
+        self.position_embedding = nn.Embedding(config.num_image_positions, config.hidden_size)
+        self.register_buffer(
+            "position_ids", torch.arange(config.num_image_positions).expand((1, -1)), persistent=False
+        )
+
+    def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
+        batch_size, patches, height, width = pixel_values.shape
+        if height != self.image_size:
+            raise ValueError(f"Input image size ({height}) doesn't match model" f" ({self.image_size}).")
+        target_dtype = self.patch_embedding.weight.dtype
+        patch_embeds = self.patch_embedding(pixel_values.to(dtype=target_dtype))
+
+        class_embeds = self.class_embedding.expand(batch_size, patches, 1, -1)
+        embeddings = torch.cat([class_embeds, patch_embeds], dim=2)
+        embeddings = embeddings + self.position_embedding(self.position_ids).unsqueeze(1)
+        return embeddings.flatten(0, 1)  # NOTE: DON'T FLATTEN MORE TO MATCH ORIG IMPL
+
+
+class MolmoVisionMLP(CLIPMLP):
+    pass
+
+
+class MolmoVisionEncoderLayer(CLIPEncoderLayer):
+    def __init__(self, config: MolmoVisionConfig):
+        super().__init__()
+        self.self_attn = MOLMO_VISION_ATTENTION_CLASSES[config._attn_implementation](config)
+        self.mlp = MolmoVisionMLP(config)
+
```

Comment:
```suggestion
```
when none are different, there is not point in redefining the layer, unless the layer is explicitly used (maybe MolmoVisionEncoderLayer)